pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,439 @@
+	return a.instance.InstanceId
+}
+
+func (a *alibabaInstance) Addresses() []string {
+	var primaryIPAddresses []string
+	for _, networkInterface := range a.instance.NetworkInterfaces.NetworkInterface {
+		primaryIPAddresses = append(primaryIPAddresses, networkInterface.PrimaryIpAddress)
+	}
+
+	return primaryIPAddresses
+}
+
+func (a *alibabaInstance) Status() instance.Status {
+	return instance.Status(a.instance.Status)
+}
+
+// New returns an Alibaba cloud provider
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	return spec, nil
+}
+
+func (p *provider) Validate(machineSpec v1alpha1.MachineSpec) error {
+	c, pc, err := p.getConfig(machineSpec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.AccessKeyID == "" {
+		return errors.New("accessKeyID is missing")
+	}
+	if c.AccessKeySecret == "" {
+		return errors.New("accessKeySecret is missing")
+	}
+	if c.RegionID == "" {
+		return errors.New("regionID is missing")
+	}
+	if c.InstanceType == "" {
+		return errors.New("instanceType is missing")
+	}
+	if c.VSwitchID == "" {
+		return errors.New("vSwitchID is missing")
+	}
+	if c.InternetMaxBandwidthOut == "" {
+		return errors.New("internetMaxBandwidthOut is missing")
+	}
+	if c.ZoneID == "" {
+		return errors.New("zoneID is missing")
+	}
+	_, err = p.getImageIDForOS(machineSpec, pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	return nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	foundInstance, err := getInstance(client, machine.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	switch instanceStatus(foundInstance.Status) {
+	case stoppedStatus:
+		startRequest := ecs.CreateStartInstanceRequest()
+		startRequest.InstanceId = foundInstance.InstanceId
+
+		_, err = client.StartInstance(startRequest)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("instance %v is in a stopped state", foundInstance.InstanceId)
+	case runningStatus:
+		if len(foundInstance.PublicIpAddress.IpAddress) == 0 {
+			ipAddress := ecs.CreateAllocatePublicIpAddressRequest()
+			ipAddress.InstanceId = foundInstance.InstanceId
+
+			_, err = client.AllocatePublicIpAddress(ipAddress)
+			if err != nil {
+				return nil, err
```

Comment:
Please wrap this and all other errors so its possible to figure out where exactly a given error comes from